File: src/test/java/com/app/StreamUtilsTest.java
package com.app;

import org.apache.kafka.clients.consumer.ConsumerRecord;
import org.apache.kafka.common.TopicPartition;
import org.apache.kafka.common.serialization.StringDeserializer;
import org.apache.spark.sql.*;
import org.apache.spark.SparkConf;
import org.apache.spark.streaming.Duration;
import org.apache.spark.streaming.api.java.JavaInputDStream;
import org.apache.spark.streaming.api.java.JavaStreamingContext;
import org.junit.After;
import org.junit.Before;
import org.junit.Rule;
import org.junit.Test;
import org.junit.rules.TemporaryFolder;

import java.io.IOException;
import java.util.HashMap;
import java.util.Map;

public class StreamUtilsTest {

    @Rule
    public TemporaryFolder tempFolder = new TemporaryFolder();

    private static String pathToCSV;
    private static String topic;
    private static Duration duration;
    private static SparkConf conf;
    private static JavaStreamingContext streamingContext;
    private static SparkSession sparkSession;
    private static Map<String, Object> kafkaParams;

    @Before
    public void initialize() {
        pathToCSV = tempFolder.getRoot().getPath();
        topic = "hotels10";
        duration = new Duration(1000);
        conf = new SparkConf().setAppName("Streaming Homework Testing").setMaster("local[*]");
        streamingContext = new JavaStreamingContext(conf, duration);
        sparkSession = SparkSession.builder().getOrCreate();
        kafkaParams = new HashMap<>();
        kafkaParams.put("bootstrap.servers", "localhost:9092");
        kafkaParams.put("key.deserializer", StringDeserializer.class);
        kafkaParams.put("value.deserializer", StringDeserializer.class);
        kafkaParams.put("group.id", "stream-hw");
        kafkaParams.put("kafka.consumer.id", "kafka-consumer-01");
        kafkaParams.put("auto.offset.reset", "latest");
        kafkaParams.put("enable.auto.commit", false);
    }

    @After
    public void after() throws IOException {
        System.out.println(pathToCSV);
        //Files.newDirectoryStream(Paths.get(pathToCSV), p -> p.toString().endsWith(".csv")).forEach(System.out::println);
    }

    @Test
    public void test() throws Exception {
//        Map<TopicPartition, Long> fromOffsets = new HashMap<>();
  //      fromOffsets.put(new TopicPartition(topic,0), 0L);
        Map<TopicPartition, Long> fromOffsets = StreamUtilsOld.getFromOffsets(topic, sparkSession, pathToCSV);
        JavaInputDStream<ConsumerRecord<String, String>> stream = StreamUtilsOld.getStream(streamingContext, topic, kafkaParams, fromOffsets);
        StreamUtilsOld.writeRDDs(stream, sparkSession, pathToCSV);

        streamingContext.start();
//        streamingContext.awaitTermination();
        streamingContext.awaitTerminationOrTimeout(2000);

        //wait(5000);
//        System.out.println(streamingContext.getState());
            //streamingContext.getState()==StreamingContextState.STOPPED

//        Files.newDirectoryStream(Paths.get(pathToCSV), p -> p.toString().endsWith(".csv")).forEach(System.out::println);
    }

}

File: src/test/scala/com/app/StreamUtilsTest.scala
package com.app

import java.nio.file.{Files, Paths}
import java.util.Properties

import org.apache.kafka.clients.admin.{AdminClient, ListTopicsOptions}
import org.apache.kafka.common.serialization.StringDeserializer
import org.apache.spark.SparkConf
import org.apache.spark.sql.SparkSession
import org.apache.spark.streaming.{Seconds, StreamingContext}
import org.scalatest.{BeforeAndAfter, FlatSpec}

import scala.reflect.io.Path
import scala.util.Try

class StreamUtilsTest extends FlatSpec with BeforeAndAfter {

  private var sparkConf: SparkConf = _
  private var ssc: StreamingContext = _
  private var sparkSession: SparkSession = _

  private val pathToCSV: String = "src/test/resources/stream"
  private val topic: String = "hotels10"

  private val kafkaParams: Map[String, Object] = Map[String, Object](
    "bootstrap.servers" -> "localhost:9092",
    "key.deserializer" -> classOf[StringDeserializer],
    "value.deserializer" -> classOf[StringDeserializer],
    "group.id" -> "stream-hw",
    "kafka.consumer.id" -> "kafka-consumer-01",
    "auto.offset.reset" -> "latest",
    "enable.auto.commit" -> (false: java.lang.Boolean)
  )

  def pathFolderExists(): Boolean = {
    Files.exists(Paths.get(pathToCSV))
  }

  before {
    val kafkaProps = new Properties()
    for ( (k,v) <- kafkaParams ) kafkaProps.put(k,v)

    try {
      AdminClient.create(kafkaProps)
          .listTopics(new ListTopicsOptions()
          .timeoutMs(3000)).listings().get()
    } catch {
      case e: Exception =>
        println("Kafka is not available: "+e.getMessage)
        println("Please install and configure it locally before testing, you can find all details in the README file.")
        sys.exit
    }

      sparkConf = new SparkConf().setMaster("local[*]").setAppName("test-streaming")
      ssc = new StreamingContext(sparkConf, Seconds(1))
      sparkSession = SparkSession.builder.getOrCreate
    }

  after {
    Thread.sleep(1000)

    ssc.stop()
    sparkSession.stop()

    Try(Path(pathToCSV).deleteRecursively)
  }

  behavior of "StreamUtils.writeRDDs method"
  it should "save records to the disk" in {

    assert(!pathFolderExists())

    val fromOffsets = StreamUtils.getFromOffsets(topic, sparkSession, pathToCSV)
    assert(fromOffsets.toList.head._2 == 0L)

    val stream = StreamUtils.getStream(ssc, topic, kafkaParams, fromOffsets)
    StreamUtils.writeRDDs(stream, sparkSession, pathToCSV)

    ssc.start()
    ssc.awaitTerminationOrTimeout(2000)

    Thread.sleep(2000)

    assert(pathFolderExists())

    val actualDF = sparkSession.read.csv(pathToCSV)
    assert(actualDF.count()==10)

    val expectedDF = sparkSession.read.csv("src/test/resources/hotels10.csv")
    assert(actualDF.collect.sameElements(expectedDF.collect))
  }

}